engine/apps/api/permissions/__init__.py

```diff
@@ -1,5 +1,196 @@
+
+    def has_object_permission(self, request: Request, view: ViewSetOrAPIView, obj: typing.Any) -> bool:
+        rbac_object_permissions: RBACObjectPermissionsAttribute = getattr(view, RBAC_OBJECT_PERMISSIONS_ATTR, None)
+
+        if rbac_object_permissions:
+            action = _get_view_action(request, view)
+
+            for PermissionClass, actions in rbac_object_permissions.items():
+                if action in actions:
+                    return PermissionClass.has_object_permission(request, view, obj)
+            return False
+
+        # has_object_permission is called after has_permission, so return True if in view there is not
+        # RBAC_OBJECT_PERMISSIONS_ATTR attr which mean no additional check involving object required
+        return True
+
+
+def has_permissions(
+    user_permissions: typing.List[GrafanaPermission], required_permissions: typing.List[RBACPermission.Permissions]
+) -> bool:
+    user_has_all_required_permissions = True
+
+    for required_permission in required_permissions:
+        granted_permission: typing.Union[None, GrafanaPermission] = None
+
+        for user_permission in user_permissions:
+            if user_permission["action"] == required_permission.value:
+                granted_permission = user_permission
+                break
+
+        if granted_permission is None:
+            user_has_all_required_permissions = False
+            break
+
+    return user_has_all_required_permissions
+
+
+class IsOwner(permissions.BasePermission):
+    def __init__(self, ownership_field: typing.Optional[str] = None) -> None:
+        self.ownership_field = ownership_field
+
+    def has_object_permission(self, request: Request, _view: ViewSet, obj: typing.Any) -> bool:
+        owner = obj if self.ownership_field is None else getattrd(obj, self.ownership_field)
+        return owner == request.user
+
+
+class HasRBACPermissions(permissions.BasePermission):
+    def __init__(self, required_permissions: typing.List[RBACPermission.Permissions]) -> None:
+        self.required_permissions = required_permissions
+
+    def has_object_permission(self, request: Request, _view: ViewSetOrAPIView, _obj: typing.Any) -> bool:
+        return has_permissions(request.user.permissions, self.required_permissions)
+
+
+class IsOwnerOrHasRBACPermissions(permissions.BasePermission):
+    def __init__(
+        self,
+        required_permissions: typing.List[RBACPermission.Permissions],
+        ownership_field: typing.Optional[str] = None,
+    ) -> None:
+        self.IsOwner = IsOwner(ownership_field)
+        self.HasRBACPermissions = HasRBACPermissions(required_permissions)
+
+    def has_object_permission(self, request: Request, view: ViewSetOrAPIView, obj: typing.Any) -> bool:
+        return self.IsOwner.has_object_permission(request, view, obj) or self.HasRBACPermissions.has_object_permission(
+            request, view, obj
+        )
+
+
+class IsStaff(permissions.BasePermission):
```

Comment:
After some trial-and-error I've decided to keep this _only_ for the `oss_installation` django app. The reason being is that without this we run into a chicken-and-egg problem.

Grafana already does have a `plugins:install` permission that would be perfect here. But since we are syncing user's permissions from Grafana to OnCall, the user's permissions would already need to be in the OnCall database.

The alternative (which I may end up refactoring to now that I think of it) would be to make a `permissions.BasePermission` subclass something like this, and replace `IsStaff` with it:
```python
class IsPluginInstaller(permissions.BasePermission):

  def has_permission(self, request: Request, _view: ViewSet) -> bool:
    # pseudocode
    return GrafanaAPIClient.user_has_plugin_install_permission(request.user)
```